customOverlay.wake

```diff
@@ -0,0 +1,80 @@
+#
+# Wake API for Specifying a custom Devicetree Overlay
+#
+# If the needs of your project aren't met by the heuristics
+# of the python-based overlay generator, you can instead
+# generate a custom overlay using these wake rules to specify
+# the properties in the chosen node.
+#
+# Example:
+# --------
+#
+# def entry = makeDevicetreeChosenMemoryEntry "/soc/bootrom@20000000" 0 0
+# def ram = makeDevicetreeChosenMemoryRam "/soc/sram@80000000" 0 0
+# def itim = None
+# def chosenNode = makeDevicetreeChosenNode entry ram itim
+#
+# makeDevicetreeCustomOverlay ("core.dts", Nil) chosenNode
+# | (writeDevicetreeCustomOverlay "design.dts" _)
+#
+
+
+tuple DevicetreeChosenMemory =
+  global Property:  String
+  global NodePath:  String
+  global RegTuple:  Integer
+  global RegOffset: Integer
+
+global def makeDevicetreeChosenMemoryEntry nodePath regTuple regOffset =
+  DevicetreeChosenMemory "metal,entry" nodePath regTuple regOffset
+
+global def makeDevicetreeChosenMemoryRam nodePath regTuple regOffset =
+  DevicetreeChosenMemory "metal,ram" nodePath regTuple regOffset
+
+global def makeDevicetreeChosenMemoryItim nodePath regTuple regOffset =
+  DevicetreeChosenMemory "metal,itim" nodePath regTuple regOffset
+
+tuple DevicetreeChosenNode =
+  global Entry: DevicetreeChosenMemory
+  global Ram:   DevicetreeChosenMemory
+  global Itim:  Option DevicetreeChosenMemory
+
+global def makeDevicetreeChosenNode entry ram itim =
+  DevicetreeChosenNode entry ram itim
+
+tuple DevicetreeCustomOverlay =
+  global Includes:           List String
+  global ChosenNode:         DevicetreeChosenNode
+  global AdditionalContents: String
+
+global def makeDevicetreeCustomOverlay includes chosenNode =
+  DevicetreeCustomOverlay includes chosenNode ""
+
+global def writeDevicetreeCustomOverlay file customOverlay =
+  def makeInclude include =
+    "/include/ \"{include}\"\n", Nil
+  def includes = mapFlat makeInclude customOverlay.getDevicetreeCustomOverlayIncludes
+
+  def chosenContents =
+    def makeLine chosenMemory =
+      def property = chosenMemory.getDevicetreeChosenMemoryProperty
+      def reference = "&\{{chosenMemory.getDevicetreeChosenMemoryNodePath}\}"
```

Comment:
FYI, Wake has a syntax for changing the symbol used for string interpolation. It's part of the multiline string syntax, but it could be useful if escaping a lot of `{` characters. See https://github.com/sifive/wake/pull/228 for the best documentation, as there's only a little bit of documentation in the Wake [quickref](https://github.com/sifive/wake/blob/master/share/doc/wake/quickref.md#primitives).

Reply:
Thanks for the heads up here. For this case I think it's more clear (especially to wake newbies) to escape the curly braces, but I'll keep it in mind for when I need a lot of escapes.